ENH: expose ellipse_phantom and utils, closes #519

adler-j opened the pull request.

kohr-h commented on odl/phantom/geometric.py:
```diff
@@ -171,23 +171,345 @@ def indicate_proj_axis(space, scale_structures=0.5):
+    length of its principial axes (axis_1, axis_2, [axis_2]) and euler angles.
+
+    This function is heavily optimized, achieving runtimes about 20 times
+    faster than "trivial" implementations. It is therefore recommended to use
```
use it

kohr-h commented:
What exactly happened in this pull? It's hard to see since stuff has been moved and modified at the same time.

Next time, please make a separate commit for the move only, then change the code. It's also much more better when trying to rebase, since git also has a hard time understanding move-and-modify commits.

adler-j commented:
> What exactly happened in this pull? It's hard to see since stuff has been moved and modified at the same time.

Basically exposed `ellipse_phantom` and added a `shepp_logan_ellipses` that can be used with it.

> Next time, please make a separate commit for the move only, then change the code. It's also much more better when trying to rebase, since git also has a hard time understanding move-and-modify commits.

Duly noted.

adler-j commented:
Will merge after CI.

kohr-h commented:
> Will merge after CI.

Not so fast, I'll have a look at the texts first.

kohr-h commented on odl/phantom/geometric.py:
```diff
@@ -171,23 +171,345 @@ def indicate_proj_axis(space, scale_structures=0.5):
+    space : `DiscreteLp`
+        Space the phantom should live in, must be 2 or 3 dimensional.
+    ellipses : `sequence` of `sequence`
+        If ``space`` is 2 dimensional each row should contain:
```
2-dimensional

kohr-h commented on odl/phantom/geometric.py:
```diff
@@ -171,23 +171,345 @@ def indicate_proj_axis(space, scale_structures=0.5):
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        Space the phantom should live in, must be 2 or 3 dimensional.
```
"live in" weird, better "be created in"
The whole sentence could be clearer, like "Space in which the phantom is created, must be 2- or 3-dimensional."

kohr-h commented on odl/phantom/geometric.py:
```diff
@@ -171,23 +171,345 @@ def indicate_proj_axis(space, scale_structures=0.5):
+
+        'value', 'axis_1', 'axis_2', 'center_x', 'center_y', 'rotation'
+
+        If space is 3 dimensional each row should contain:
```
3-dimensional

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -22,14 +22,17 @@
-    """Return ellipse parameters for a 2d Shepp-Logan phantom."""
+    """Return ellipse parameters for a 2d Shepp-Logan phantom.
+
+    This assumes that the ellipses are on the square [-1, -1]x[-1, -1].
```
"are on" -> "are contained in"

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -44,7 +47,10 @@ def _shepp_logan_ellipse_2d():
-    """Return ellipse parameters for a 3d Shepp-Logan phantom."""
+    """Return ellipse parameters for a 3d Shepp-Logan phantom.
+
+    This assumes that the ellipses are on the cube [-1, -1, -1]x[1, 1, 1].
```
"are contained in"

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -65,12 +71,14 @@ def _modified_shepp_logan_ellipses(ellipses):
     """
     intensities = [1.0, -0.8, -0.2, -0.2, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
 
+    assert len(ellipses) == len(intensities)
```
Should be a proper error message.

adler-j replied:
internal helpers can do asserts, no? I thought that was policy.

kohr-h replied:
I'm not so sure about that. To me, it depends on the nature of the internal helper. For instance, if the public function delegates everything, including checks, to internal helpers, I'm against asserts since they provide users with very little helpful information. They're not so internal in such cases, after all, and to me, this is such a case.

I'm okay with asserts when a deviation from the check is a bug rather than the result of invalid input. Internal consistency checks can use asserts, input validation should not in my opinion.

adler-j replied:
Well this function should not be called by any external user, ever. The only calls come through from `shepp_logan_ellipses`, and this check is to make sure that the number of ellipses in those are OK, which they currently are.

The only way to make this assert fail is by either calling a private method or by editing the ODL source code. To me that is a typical case for asserts?

kohr-h replied:
So there's  no way for a user to provide own ellipses? In that case it's OK.

adler-j replied:
Not to this one, it's basically a utility so we wont have to write two functions that create the modified ellipses. If a user wants to use this the correct usage would be something like

```python
ellipses = shepp_logan_ellipses(3, modified=True)
# modify ellipses however you want
```

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -95,6 +108,27 @@ def shepp_logan(space, modified=False):
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        The 2/3 dimension space that the phantom should be created in.
```
2- or 3-dimensional

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -95,6 +108,27 @@ def shepp_logan(space, modified=False):
+        The 2/3 dimension space that the phantom should be created in.
+    modified : `bool`, optional
+        True if the modified Shepp-Logan phantom should be given.
+        The modified phantom has greatly amplified contrast to aid in
```
aid ~~in~~

kohr-h commented on odl/phantom/transmission.py:
```diff
@@ -95,6 +108,27 @@ def shepp_logan(space, modified=False):
+
+    References
+    ----------
+    .. Shepp-Logan phantom: en.wikipedia.org/wiki/Shepp–Logan_phantom
```
`.. _Shepp-Logan phantom` (underscore missing)

kohr-h commented:
Some minor stuff, after that you can go ahead :-)

kohr-h commented:
This should also be mentioned in the release notes, BTW.

adler-j commented:
> This should also be mentioned in the release notes, BTW.

It is: 

Expose ``ellipse_phantom`` and ``shepp_logan_ellipses`` to ``odl.phantom``
